src/Security/Authentication/Core/src/JsonDocumentAuthExtensions.cs

```diff
@@ -0,0 +1,14 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text.Json;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    public static class JsonDocumentAuthExtensions
```

Comment:
This extension looks so basic that it seems strange it's not directly part of `JsonDocument`. Is such an API planned?

Reply:
Yes, we use it across multiple product assemblies, samples, tests, etc., and it's something our customers will want for easy migration.

Reply:
I did keep the namespace constrained to only auth users though.

Reply:
It feels like the wrong place for the extension to exists. If it's general purpose enough that your code and user code would benefit from it, we should move it down-level. @ahsonkhan could we consider adding this as an extension in System.Text.Json?

Reply:
See my comment here: https://github.com/aspnet/AspNetCore/pull/7105#discussion_r252091383
I am not sure why such an extension method exists. And if the goal is primarily to create a custom "null" string behavior, it doesn't make sense to provide an in-box implementation for that.

`JsonElement.GetString()` already returns null when the token type is null.

Reply:
It's not special casing null strings, it's special casing missing properties and avoids the `KeyNotFoundException : The given key was not present in the dictionary.` thrown by GetProperty for missing elements. The extension is used 8 times in product code and 5 in samples, and I expect 3rd party OAuth implementations to use it as well. @ahsonkhan I can see why you wouldn't add this directly to JsonDocument, but it suits our scenarios well.

Consider the example here:
https://github.com/aspnet/AspNetCore/pull/7105/files#diff-a3f5673336ed9c217610a732aec6cdcfL14
This extension gives us a drop in replacement for prior functionality. The TryGetProperty pattern encapsulated by the extension is x3 as long.

Reply:
> This extension gives us a drop in replacement for prior functionality.

But will likely result in quite slow perf since the extension is on `JsonDocument`.

In your example, it might make more sense for the `response` to be a `JsonElement`, not `JsonDocument`.

```C#
// Before:
AccessToken = response.Value<string>("access_token");
```

```C#
// After:
AccessToken = response.GetString("access_token");
```

I understand the usability concern. We currently do not have an API that searches the `JsonElement` for a property and returns the .NET type value back to you in one call. **We can consider adding an API similar to `Value<T>` to help alleviate some of the concern here. Thanks for the feedback :)**

I would appreciate it if you can file an issue in corefx to track this. Otherwise, I will do so.

One thing we should consider is making sure we are consistent around "null" or key not found behavior.

```C#
var sample = JObject.Parse("{\"message\": 42}");
JToken token = sample["foo"]; // Why shouldn't this throw?
string value = sample.Value<string>("foo"); // And if the above throws, this returning null might be confusing
```

Reply:
Filed https://github.com/dotnet/corefx/issues/34997

Note AspNetCore has banned indexers that throw key not found exceptions, they're a significant usability barrier. All indexers return null or defaults for missing values. If that's not appropriate for a type then that type does not get an indexer.